include/boost/parameter/aux_/tagged_argument.hpp

```diff
@@ -30,9 +39,33 @@ struct tagged_argument_base {};
 template <class Keyword, class Arg>
 struct tagged_argument : tagged_argument_base
 {
+    typedef typename ::boost::remove_const<Arg>::type arg_type;
+
+public:
     typedef Keyword key_type;
-    typedef Arg value_type;
-    typedef Arg& reference;
+
+    // Wrap plain (non-UDT) function objects in either
+    // a boost::function or a std::function. -- Cromwell D. Enage
+    typedef typename ::boost::mpl::if_<
+        ::boost::is_function<arg_type>
+#if defined(BOOST_NO_CXX11_HDR_FUNCTIONAL)
```

Comment:
If you make msvc-11.0 use boost::function here, does the need to have workarounds when using it disappear?

Reply:
I just tried it on my development branch.  Unfortunately, I still need the workarounds in the test program, or it won't work for msvc-11.0.

Reply:
I'm still attempting to make the workarounds for msvc-11.0 unnecessary in user code.  I'm focusing my efforts on boost::parameter::keyword.  My goal is to implement an additional overloaded assignment operator that wraps the tag metafunction return type around boost::function wrapped around the argument, then use boost::enable_if et. al. with boost::is_function to avoid compiler errors mentioning ambiguous calls.  Here are my issues:

1. If both overloads take in T const&, then compose.cpp compiles and links but fails to execute because the compiler never resolves the call with the function-accepting overload.
2. If the function-accepting overload takes in T while the other overload takes in T const&, then I still encounter the ambiguous-call compiler error because I haven't figured out how to use boost::lazy_disable_if with boost::is_function for the other overload.
3. If I convert the assignment operator overloads to overloads of a static helper function and use boost::declval with either decltype or BOOST_TYPEOF_TPL to compute the return type of the driver assignment operator, then compose.cpp and other test programs fail to compile because the static helper function cannot be found.

If there are more promising approaches that I'm overlooking, please let me know.

Reply:
I'm still attempting to make the workarounds for msvc-11.0 unnecessary in user code.  I'm focusing my efforts on boost::parameter::keyword.  My goal is to implement an additional overloaded assignment operator that wraps the tag metafunction return type around boost::function wrapped around the argument, then use boost::enable_if et. al. with boost::is_function to avoid compiler errors mentioning ambiguous calls.  Here are my issues:

1. If both overloads take in T const&, then compose.cpp compiles and links but fails to execute because the compiler never resolves the call with the function-accepting overload.
2. If the function-accepting overload takes in T while the other overload takes in T const&, then I still encounter the ambiguous-call compiler error because I haven't figured out how to use boost::lazy_disable_if with boost::is_function for the other overload.
3. If I convert the assignment operator overloads to overloads of a static helper function and use boost::declval with either decltype or BOOST_TYPEOF_TPL to compute the return type of the driver assignment operator, then compose.cpp and other test programs fail to compile because the static helper function cannot be found.

If there are more promising approaches that I'm overlooking, please let me know.

Reply:
I'm still attempting to make the workarounds for msvc-11.0 unnecessary in user code.  I'm focusing my efforts on boost::parameter::keyword.  My goal is to implement an additional overloaded assignment operator that wraps the tag metafunction return type around boost::function wrapped around the argument, then use boost::enable_if et. al. with boost::is_function to avoid compiler errors mentioning ambiguous calls.  Here are my issues:

1. If both overloads take in T const&, then compose.cpp compiles and links but fails to execute because the compiler never resolves the call with the function-accepting overload.
2. If the function-accepting overload takes in T while the other overload takes in T const&, then I still encounter the ambiguous-call compiler error because I haven't figured out how to use boost::lazy_disable_if with boost::is_function for the other overload.
3. If I convert the assignment operator overloads to overloads of a static helper function and use boost::declval with either decltype or BOOST_TYPEOF_TPL to compute the return type of the driver assignment operator, then compose.cpp and other test programs fail to compile because the static helper function cannot be found.

If there are more promising approaches that I'm overlooking, please let me know.

Reply:
I'm still attempting to make the workarounds for msvc-11.0 unnecessary in user code.  I'm focusing my efforts on boost::parameter::keyword.  My goal is to implement an additional overloaded assignment operator that wraps the tag metafunction return type around boost::function wrapped around the argument, then use boost::enable_if et. al. with boost::is_function to avoid compiler errors mentioning ambiguous calls.  Here are my issues:

1. If both overloads take in T const&, then compose.cpp compiles and links but fails to execute because the compiler never resolves the call with the function-accepting overload.
2. If the function-accepting overload takes in T while the other overload takes in T const&, then I still encounter the ambiguous-call compiler error because I haven't figured out how to use boost::lazy_disable_if with boost::is_function for the other overload.
3. If I convert the assignment operator overloads to overloads of a static helper function and use boost::declval with either decltype or BOOST_TYPEOF_TPL to compute the return type of the driver assignment operator, then compose.cpp and other test programs fail to compile because the static helper function cannot be found.

If there are more promising approaches that I'm overlooking, please let me know.